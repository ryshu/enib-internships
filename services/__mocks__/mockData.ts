export function defaultBusiness() {
    const VALID_BUSINESS: IBusinessEntity = {
        name: 'test',
        country: 'France',
        city: 'Brest',
        postalCode: '29200',
        address: 'TEST',
    };
    return VALID_BUSINESS;
}

export function defaultFiles() {
    const VALID_FILE: IFileEntity = {
        name: 'fichier',
        size: 150,
        type: 'john.doe@enib.fr',
        path: 'S10',
    };
    return VALID_FILE;
}

export function defaultInternships() {
    const VALID_INTERNSHIP: IInternshipEntity = {
        subject: 'test',
        description: 'Stage',
        country: 'France',
        city: 'Brest',
        postalCode: '29280',
        address: 'TEST',
        isLanguageCourse: true,
        isValidated: false,
    };
    return VALID_INTERNSHIP;
}

export function defaultMentoringPropositions() {
    const VALID_MENTORING_PROPOSITION: IMentoringPropositionEntity = {
        comment: 'test',
    };
    return VALID_MENTORING_PROPOSITION;
}

export function defaultStudents() {
    const VALID_STUDENT: IStudentEntity = {
        firstName: 'John',
        lastName: 'Doe',
        email: 'john.doe@enib.fr',
        semester: 'S10',
    };
    return VALID_STUDENT;
}

export function defaultInternshipTypes() {
    const VALID_INTERNSHIP_TYPES: IInternshipTypeEntity = {
        label: 'test',
    };
    return VALID_INTERNSHIP_TYPES;
}
