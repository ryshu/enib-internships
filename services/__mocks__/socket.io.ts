import SubSocket from './subsocket';

export class SocketServer {
    public of(_nsp: string | RegExp): any {
        return new SubSocket();
    }

    public close(fn?: () => void): void {
        fn();
    }

    public on(_event: 'connection', _listener: (socket: any) => void): any {
        return;
    }

    public emit(_event: string, ..._args: any[]): any {
        return jest.fn();
    }

    public to() {
        return this;
    }
}

export default function() {
    return new SocketServer();
}
