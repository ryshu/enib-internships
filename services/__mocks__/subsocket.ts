export default class SubSocket {
    public emit ( ..._args: any[]): void {
        return;
    }
}
