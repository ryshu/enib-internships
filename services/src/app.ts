import express, { Request, Response, NextFunction } from 'express';
import bodyParser from 'body-parser';
import cors from 'cors';
import compression from 'compression'; // compresses requests
import lusca from 'lusca';
import dotenv from 'dotenv';
import flash from 'express-flash';
import path from 'path';
import methodOverride from 'method-override';

// Load environment variables from .env file, where API keys and passwords are configured
dotenv.config({ path: '.env' });

// Loggers
import logger from './utils/logger';
import { APIError } from './utils/error';
import httpStatus from 'http-status-codes';

// router and database
import router from './configs/setup/route';
import './configs/instances/database'; // Only import to setup
import './configs/setup/database'; // Only import to setup

// Create Express server
const app = express();

// Express configuration
app.set('port', process.env.INTERNSHIP_ENIB_API_PORT || 3000);
app.use(compression());
app.use(bodyParser.json({ limit: '1mb' }));
app.use(bodyParser.urlencoded({ extended: true, limit: '1mb' }));
app.use(flash());
app.use(lusca.xframe('SAMEORIGIN'));
app.use(lusca.xssProtection(true));

app.use(express.static(path.join(__dirname, 'public'), { maxAge: 31557600000 }));

const corsOptions = {
    origin: 'http://localhost:4200',
    optionsSuccessStatus: 200,
};
app.use(cors(corsOptions));

app.use(router);

/**
 * Error handler setup
 */
app.use(methodOverride());
app.use((err: APIError | Error, _req: Request, res: Response, _next: NextFunction) => {
    if (err instanceof APIError) {
        return res.status(err.status).json(err);
    } else {
        logger.error(err);
        return res.status(httpStatus.INTERNAL_SERVER_ERROR).send(`Internal server error`);
    }
});

export default app;
