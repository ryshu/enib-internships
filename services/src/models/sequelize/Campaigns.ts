import * as Sequelize from 'sequelize';

import database from '../../configs/instances/database';

import MPS from './MentoringPropositions';
import Mentors from './Mentors';
import InternshipTypes from './InternshipTypes';
import Internships from './Internships';

import {
    IInternshipEntity,
    IMentoringPropositionEntity,
    IInternshipTypeEntity,
} from '../../declarations';

class Campaigns extends Sequelize.Model {
    public static associations: {
        propositions: Sequelize.Association<Campaigns, MPS>;
        availbleInternships: Sequelize.Association<Campaigns, Internships>;
        validatedInternships: Sequelize.Association<Campaigns, Internships>;
        mentors: Sequelize.Association<Campaigns, Mentors>;
        category: Sequelize.Association<Campaigns, InternshipTypes>;
    };

    public id!: number; // Note that the `null assertion` `!` is required in strict mode.

    public name: string;
    public description: string;

    public semester: string;
    public maxProposition: number;

    public isPublish: boolean;

    public startAt: number;
    public endAt: number;

    // timestamps!
    public readonly createdAt!: Date;
    public readonly updatedAt!: Date;

    // Proposition
    public getPropositions: Sequelize.HasManyGetAssociationsMixin<MPS>;
    public addProposition: Sequelize.HasManyAddAssociationMixin<MPS, MPS['id']>;
    public createProposition: Sequelize.HasManyCreateAssociationMixin<IMentoringPropositionEntity>;
    public hasProposition: Sequelize.HasManyHasAssociationMixin<MPS, MPS['id']>;
    public countPropositions: Sequelize.HasManyCountAssociationsMixin;

    // Mentor
    public getMentors: Sequelize.BelongsToManyGetAssociationsMixin<Mentors>;
    public addMentor: Sequelize.BelongsToManyAddAssociationMixin<Mentors, Mentors['id']>;
    public hasMentor: Sequelize.BelongsToManyHasAssociationMixin<Mentors, Mentors['id']>;
    public countMentors: Sequelize.BelongsToManyCountAssociationsMixin;

    // Category
    public getCategory: Sequelize.BelongsToGetAssociationMixin<InternshipTypes>;
    public setCategory: Sequelize.BelongsToSetAssociationMixin<
        InternshipTypes,
        InternshipTypes['id']
    >;
    public createCategory: Sequelize.BelongsToCreateAssociationMixin<IInternshipTypeEntity>;

    // AvailableInternships
    public getAvailableInternships: Sequelize.HasManyGetAssociationsMixin<Internships>;
    public addAvailableInternship: Sequelize.HasManyAddAssociationMixin<
        Internships,
        Internships['id']
    >;
    public createAvailableInternship: Sequelize.HasManyCreateAssociationMixin<IInternshipEntity>;
    public hasAvailableInternship: Sequelize.HasManyHasAssociationMixin<
        Internships,
        Internships['id']
    >;
    public countAvailableInternships: Sequelize.HasManyCountAssociationsMixin;

    // ValidatedInternships
    public getValidatedInternships: Sequelize.HasManyGetAssociationsMixin<Internships>;
    public addValidatedInternship: Sequelize.HasManyAddAssociationMixin<
        Internships,
        Internships['id']
    >;
    public createValidatedInternship: Sequelize.HasManyCreateAssociationMixin<IInternshipEntity>;
    public hasValidatedInternship: Sequelize.HasManyHasAssociationMixin<
        Internships,
        Internships['id']
    >;
    public countValidatedInternships: Sequelize.HasManyCountAssociationsMixin;

    public readonly propositions?: MPS[];
    public readonly availableInternships?: Internships[];
    public readonly validatedInternships?: Internships[];
    public readonly mentors?: Mentors[];
    public readonly category: InternshipTypes;
}

Campaigns.init(
    {
        id: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            autoIncrement: true,
            primaryKey: true,
        },
        name: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        description: {
            type: new Sequelize.DataTypes.TEXT(),
            allowNull: false,
        },

        semester: {
            type: Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        maxProposition: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            defaultValue: 0,
            allowNull: true,
        },

        isPublish: {
            type: Sequelize.DataTypes.BOOLEAN,
            defaultValue: false,
        },

        startAt: {
            type: Sequelize.DataTypes.BIGINT,
            allowNull: true,
            defaultValue: null,
        },
        endAt: {
            type: Sequelize.DataTypes.BIGINT,
            allowNull: true,
            defaultValue: null,
        },
    },
    {
        tableName: 'campaigns',
        sequelize: database,
        defaultScope: {
            attributes: { exclude: ['categoryId'] },
        },
    },
);

export default Campaigns;
