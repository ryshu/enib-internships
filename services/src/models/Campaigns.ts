import * as Sequelize from 'sequelize';

import database from '../configs/instances/database';

import MPS from './MentoringPropositions';
import Internships from './Internships';

class Campaigns extends Sequelize.Model implements ICampaignEntity {
    public static associations: {
        propositions: Sequelize.Association<Campaigns, MPS>;
        availbleInternships: Sequelize.Association<Campaigns, Internships>;
        validatedInternships: Sequelize.Association<Campaigns, Internships>;
    };

    public id!: number; // Note that the `null assertion` `!` is required in strict mode.

    public name: string;
    public description: string;
    public startAt: number;
    public endAt: number;
    public semester: string;
    public maxProposition: number;

    // timestamps!
    public readonly createdAt!: Date;
    public readonly updatedAt!: Date;

    public getPropositions: Sequelize.HasManyGetAssociationsMixin<MPS>;
    public addProposition: Sequelize.HasManyAddAssociationMixin<MPS, MPS['id']>;
    public createProposition: Sequelize.HasManyCreateAssociationMixin<IMentoringPropositionEntity>;
    public hasProposition: Sequelize.HasManyHasAssociationMixin<MPS, MPS['id']>;
    public countPropositions: Sequelize.HasManyCountAssociationsMixin;

    // AvailableInternships
    public getAvailableInternships: Sequelize.HasManyGetAssociationsMixin<Internships>;
    public addAvailableInternship: Sequelize.HasManyAddAssociationMixin<Internships, Internships['id']>;
    public createAvailableInternship: Sequelize.HasManyCreateAssociationMixin<IInternshipEntity>;
    public hasAvailableInternship: Sequelize.HasManyHasAssociationMixin<Internships, Internships['id']>;
    public countAvailableInternships: Sequelize.HasManyCountAssociationsMixin;

    // ValidatedInternships
    public getValidatedInternships: Sequelize.HasManyGetAssociationsMixin<Internships>;
    public addValidatedInternship: Sequelize.HasManyAddAssociationMixin<Internships, Internships['id']>;
    public createValidatedInternship: Sequelize.HasManyCreateAssociationMixin<IInternshipEntity>;
    public hasValidatedInternship: Sequelize.HasManyHasAssociationMixin<Internships, Internships['id']>;
    public countValidatedInternships: Sequelize.HasManyCountAssociationsMixin;

    public readonly propositions?: MPS[];
    public readonly availableInternships?: Internships[];
    public readonly validatedInternships?: Internships[];
}

Campaigns.init(
    {
        id: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            autoIncrement: true,
            primaryKey: true,
        },
        name: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        description: {
            type: new Sequelize.DataTypes.STRING(),
            allowNull: false,
        },
        startAt: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            allowNull: true,
            defaultValue: null,
        },
        endAt: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            allowNull: true,
            defaultValue: null,
        },
        semester: {
            type: Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        maxProposition: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            defaultValue: 0,
            allowNull: true,
        },
    },
    {
        tableName: 'campaigns',
        sequelize: database,
    },
);

export default Campaigns;
