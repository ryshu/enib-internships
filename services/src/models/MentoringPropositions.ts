import * as Sequelize from 'sequelize';

import database from '../configs/instances/database';

import Campaigns from './Campaigns';

class MentoringPropositions extends Sequelize.Model implements IMentoringPropositionEntity {
    public static associations: {
        campaign: Sequelize.Association<MentoringPropositions, Campaigns>;
    };

    public id!: number; // Note that the `null assertion` `!` is required in strict mode.

    public comment: string;

    // timestamps!
    public readonly createdAt!: Date;
    public readonly updatedAt!: Date;

    public getCampaign: Sequelize.BelongsToGetAssociationMixin<Campaigns>;
    public setCampaign: Sequelize.BelongsToSetAssociationMixin<Campaigns, Campaigns['id']>;
    public createCampaign: Sequelize.BelongsToCreateAssociationMixin<ICampaignEntity>;

    public readonly campaign?: Campaigns | Campaigns['id'];
}

MentoringPropositions.init(
    {
        id: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            autoIncrement: true,
            primaryKey: true,
        },
        comment: {
            type: new Sequelize.DataTypes.TEXT(),
            allowNull: true,
        },
    },
    {
        tableName: 'mentoring-propositions',
        sequelize: database,
        defaultScope: {
            attributes: { exclude: ['campaignId'] },
        },
    },
);

export default MentoringPropositions;
