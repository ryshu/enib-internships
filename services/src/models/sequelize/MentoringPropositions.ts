import * as Sequelize from 'sequelize';

import database from '../../configs/instances/database';

import Campaigns from './Campaigns';
import Mentors from './Mentors';
import Internships from './Internships';

import { ICampaignEntity, IInternshipEntity, IMentorEntity } from '../../declarations';

class MentoringPropositions extends Sequelize.Model {
    public static associations: {
        mentor: Sequelize.Association<MentoringPropositions, Mentors>;
        campaign: Sequelize.Association<MentoringPropositions, Campaigns>;
        internship: Sequelize.Association<MentoringPropositions, Internships>;
    };

    public id!: number; // Note that the `null assertion` `!` is required in strict mode.

    public comment: string;

    // timestamps!
    public readonly createdAt!: Date;
    public readonly updatedAt!: Date;

    // Campaign
    public getCampaign: Sequelize.BelongsToGetAssociationMixin<Campaigns>;
    public setCampaign: Sequelize.BelongsToSetAssociationMixin<Campaigns, Campaigns['id']>;
    public createCampaign: Sequelize.BelongsToCreateAssociationMixin<ICampaignEntity>;

    // Mentors
    public getMentor: Sequelize.BelongsToGetAssociationMixin<Mentors>;
    public setMentor: Sequelize.BelongsToSetAssociationMixin<Mentors, Mentors['id']>;
    public createMentor: Sequelize.BelongsToCreateAssociationMixin<IMentorEntity>;

    // Internships
    public getInternship: Sequelize.BelongsToGetAssociationMixin<Internships>;
    public setInternship: Sequelize.BelongsToSetAssociationMixin<Internships, Internships['id']>;
    public createInternship: Sequelize.BelongsToCreateAssociationMixin<IInternshipEntity>;

    public readonly campaign?: Campaigns | Campaigns['id'];
    public readonly mentor?: Mentors | Mentors['id'];
    public readonly internship?: Internships | Internships['id'];
}

MentoringPropositions.init(
    {
        id: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            autoIncrement: true,
            primaryKey: true,
        },
        comment: {
            type: new Sequelize.DataTypes.TEXT(),
            allowNull: true,
        },
    },
    {
        tableName: 'mentoring-propositions',
        sequelize: database,
        timestamps: true,
        paranoid: true,
        defaultScope: {
            attributes: { exclude: ['campaignId', 'mentorId', 'internshipId'] },
        },
    },
);

export default MentoringPropositions;
