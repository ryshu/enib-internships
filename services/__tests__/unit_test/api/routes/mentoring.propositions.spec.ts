import request from 'supertest';

// Include app for super agent, not server
import app from '../../../../src/app';

// Include database connection
import dbSetup from '../../../../src/configs/setup/database';
import Campaigns from '../../../../src/models/Campaigns';

// Import model for pre-operation before asserting API methods
import MentoringPropositions from '../../../../src/models/MentoringPropositions';

import { defaultMentoringPropositions, defaultCampaigns } from '../../../../__mocks__/mockData';

beforeAll((done) => {
    dbSetup.then(() => done()).catch((e) => done(e));
});

describe('GET /mentoringPropositions', () => {
    beforeEach(async () => {
        // Remove all
        await MentoringPropositions.destroy({ where: {} });
    });

    it('NoMentoringPropositions_204', async () => {
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions`,
        );
        expect(RESPONSE.status).toBe(204);
    });

    it('BadRequest_400', async () => {
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions?limit=Nan`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('MentoringPropositions_200', async () => {
        const VALID_MENTORING_PROPOSITION = defaultMentoringPropositions();

        await MentoringPropositions.create(VALID_MENTORING_PROPOSITION);
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions`,
        );
        expect(RESPONSE.status).toBe(200);
        expect(Array.isArray(RESPONSE.body.data)).toBeTruthy();
        expect(RESPONSE.body).toMatchSnapshot({
            data: [
                {
                    createdAt: expect.any(String),
                    updatedAt: expect.any(String),
                },
            ],
        });
    });
});

describe('POST /mentoringPropositions', () => {
    it('NoBody_400', async () => {
        const RESPONSE = await request(app).post(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('ValidMentoringPropositions_200', async () => {
        const VALID_MENTORING_PROPOSITION = defaultMentoringPropositions();

        const RESPONSE = await request(app)
            .post(`/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions`)
            .send(VALID_MENTORING_PROPOSITION);
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            id: expect.any(Number),
        });
    });
});

describe('GET /mentoringPropositions/:id', () => {
    beforeEach(async () => {
        // Remove all
        await MentoringPropositions.destroy({ where: {} });
    });

    it('NoMentoringPropositions_204', async () => {
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions/10`,
        );
        expect(RESPONSE.status).toBe(204);
    });

    it('BadRequest_400', async () => {
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions/{falseEncoding}`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('MentoringPropositions_200', async () => {
        const VALID_MENTORING_PROPOSITION = defaultMentoringPropositions();

        const CREATED = await MentoringPropositions.create(VALID_MENTORING_PROPOSITION);
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions/${CREATED.id}`,
        );
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            id: expect.any(Number),
        });
    });
});

describe('PUT /mentoringPropositions/:id', () => {
    beforeEach(async () => {
        // Remove all
        await MentoringPropositions.destroy({ where: {} });
    });

    it('NoMentoringPropositions_204', async () => {
        const RESPONSE = await request(app).put(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions/10`,
        );
        expect(RESPONSE.status).toBe(204);
    });

    it('BadRequest_400', async () => {
        const RESPONSE = await request(app).put(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions/{falseEncoding}`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('MentoringPropositions_200_UpdateAllData', async () => {
        const VALID_MENTORING_PROPOSITION = defaultMentoringPropositions();

        const CREATED = await MentoringPropositions.create(VALID_MENTORING_PROPOSITION);

        // Change some data on VALID_MENTORING_PROPOSITION
        VALID_MENTORING_PROPOSITION.comment = 'additional data';

        const RESPONSE = await request(app)
            .put(
                `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions/${CREATED.id}`,
            )
            .send(VALID_MENTORING_PROPOSITION);
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            id: expect.any(Number),
        });
    });

    it('MentoringPropositions_200_NotAnyDataUpdated', async () => {
        const VALID_MENTORING_PROPOSITION = defaultMentoringPropositions();

        const CREATED = await MentoringPropositions.create(VALID_MENTORING_PROPOSITION);

        const RESPONSE = await request(app)
            .put(
                `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions/${CREATED.id}`,
            )
            .send({});
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            id: expect.any(Number),
        });
    });
});

describe('DELETE /mentoringPropositions/:id', () => {
    beforeEach(async () => {
        // Remove all
        await MentoringPropositions.destroy({ where: {} });
    });

    it('NoMentoringPropositions_200', async () => {
        const RESPONSE = await request(app).delete(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions/10`,
        );
        expect(RESPONSE.status).toBe(200);
    });

    it('BadRequest_400', async () => {
        const RESPONSE = await request(app).delete(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions/{falseEncoding}`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('MentoringPropositions_200', async () => {
        const VALID_MENTORING_PROPOSITION = defaultMentoringPropositions();

        const CREATED = await MentoringPropositions.create(VALID_MENTORING_PROPOSITION);

        const RESPONSE = await request(app).delete(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions/${CREATED.id}`,
        );
        expect(RESPONSE.status).toBe(200);
    });
});

describe('GET /mentoringPropositions/:id/campaigns', () => {
    beforeEach(async () => {
        // Remove all
        await MentoringPropositions.destroy({ where: {} });
    });

    it('NoInternship_204', async () => {
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions/10/campaigns`,
        );
        expect(RESPONSE.status).toBe(204);
    });

    it('BadRequest_400', async () => {
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions/{falseEncoding}/campaigns`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('MentoringPropositions_200_NoLinkedData', async () => {
        const VALID_MENTORING_PROPOSITIONS = defaultMentoringPropositions();

        const CREATED = await MentoringPropositions.create(VALID_MENTORING_PROPOSITIONS);
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions/${CREATED.id}/campaigns`,
        );
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toEqual({});
    });

    it('MentoringPropositions_200_WithLinkedData', async () => {
        const VALID_CAMPAIGN = defaultCampaigns();
        const VALID_MENTORING_PROPOSITIONS = defaultMentoringPropositions();

        const CREATED_CAMPAIGN = await Campaigns.create(VALID_CAMPAIGN);
        let CREATED_MENTORING_PROPOSITION = await MentoringPropositions.create(
            VALID_MENTORING_PROPOSITIONS,
        );

        await CREATED_MENTORING_PROPOSITION.setCampaign(CREATED_CAMPAIGN.id);
        CREATED_MENTORING_PROPOSITION = await MentoringPropositions.findByPk(
            CREATED_MENTORING_PROPOSITION.id,
        );

        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions/${CREATED_MENTORING_PROPOSITION.id}/campaigns`,
        );
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
        });
    });
});

describe('POST /mentoringPropositions/:id/campaigns/:internship_type_id/link', () => {
    beforeEach(async () => {
        // Remove all
        await MentoringPropositions.destroy({ where: {} });
    });

    it('NoInternship_204', async () => {
        const RESPONSE = await request(app).post(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions/10/campaigns/20/link`,
        );
        expect(RESPONSE.status).toBe(204);
    });

    it('BadRequest_400_WrongID', async () => {
        const RESPONSE = await request(app).post(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions/{falseEncoding}/campaigns/10/link`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('BadRequest_400_WrongBusinessID', async () => {
        const RESPONSE = await request(app).post(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions/10/campaigns/{falseEncoding}/link`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('MentoringPropositions_204_NoBusiness', async () => {
        // In this case, we check if link a existing mentoringPropositions and an unexisting campaigns work
        const VALID_MENTORING_PROPOSITIONS = defaultMentoringPropositions();

        const CREATED = await MentoringPropositions.create(VALID_MENTORING_PROPOSITIONS);
        const RESPONSE = await request(app).post(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions/${CREATED.id}/campaigns/20/link`,
        );
        expect(RESPONSE.status).toBe(204);
    });

    it('MentoringPropositions_200_WithBusiness', async () => {
        const VALID_CAMPAIGN = defaultCampaigns();
        const VALID_MENTORING_PROPOSITIONS = defaultMentoringPropositions();

        const CREATED_CAMPAIGN = await Campaigns.create(VALID_CAMPAIGN);
        let CREATED_MENTORING_PROPOSITION = await MentoringPropositions.create(
            VALID_MENTORING_PROPOSITIONS,
        );

        const RESPONSE = await request(app).post(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentoringPropositions/${CREATED_MENTORING_PROPOSITION.id}/campaigns/${CREATED_CAMPAIGN.id}/link`,
        );

        // Should answer 200
        expect(RESPONSE.status).toBe(200);

        // check if campaign and internship are linked
        CREATED_MENTORING_PROPOSITION = await MentoringPropositions.findByPk(
            CREATED_MENTORING_PROPOSITION.id,
            {
                include: [{ model: Campaigns, as: 'campaign' }],
            },
        );

        const data = JSON.parse(JSON.stringify(CREATED_MENTORING_PROPOSITION.campaign)) as any;

        expect(CREATED_MENTORING_PROPOSITION.campaign).toBeTruthy();
        expect(data).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
        });
    });
});
