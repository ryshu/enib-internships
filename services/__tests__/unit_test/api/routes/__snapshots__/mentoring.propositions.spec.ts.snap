// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /mentoringPropositions MentoringPropositions_200 1`] = `
Object {
  "data": Array [
    Object {
      "comment": "test",
      "createdAt": Any<String>,
      "id": 1,
      "updatedAt": Any<String>,
    },
  ],
  "length": 1,
  "max": 1,
  "page": 1,
}
`;

exports[`GET /mentoringPropositions/:id MentoringPropositions_200 1`] = `
Object {
  "campaign": null,
  "comment": "test",
  "createdAt": Any<String>,
  "id": Any<Number>,
  "internship": null,
  "mentor": null,
  "updatedAt": Any<String>,
}
`;

exports[`GET /mentoringPropositions/:id/campaigns MentoringPropositions_200_WithLinkedData 1`] = `
Object {
  "createdAt": Any<String>,
  "description": "A description to test",
  "endAt": 20191012,
  "id": 1,
  "isPublish": true,
  "maxProposition": 2,
  "name": "test",
  "semester": "S5",
  "startAt": 0,
  "updatedAt": Any<String>,
}
`;

exports[`GET /mentoringPropositions/:id/internships MentoringPropositions_200_WithLinkedData 1`] = `
Object {
  "additional": null,
  "address": "TEST",
  "city": "Brest",
  "country": "France",
  "createdAt": Any<String>,
  "description": "Stage",
  "endAt": 20191012,
  "id": 1,
  "isInternshipAbroad": true,
  "isProposition": false,
  "isPublish": true,
  "isValidated": false,
  "postalCode": "29280",
  "publishAt": 20191012,
  "startAt": 20191012,
  "state": "available",
  "subject": "test",
  "updatedAt": Any<String>,
}
`;

exports[`GET /mentoringPropositions/:id/mentors MentoringPropositions_200_WithLinkedData 1`] = `
Object {
  "createdAt": Any<String>,
  "email": "john.doe@enib.fr",
  "firstName": "John",
  "id": 2,
  "lastName": "Doe",
  "role": "default",
  "updatedAt": Any<String>,
}
`;

exports[`POST /mentoringPropositions ValidMentoringPropositions_200 1`] = `
Object {
  "comment": "test",
  "createdAt": Any<String>,
  "id": Any<Number>,
  "updatedAt": Any<String>,
}
`;

exports[`POST /mentoringPropositions/:id/campaigns/:internship_type_id/link MentoringPropositions_200_WithBusiness 1`] = `
Object {
  "createdAt": Any<String>,
  "description": "A description to test",
  "endAt": 20191012,
  "id": 2,
  "isPublish": true,
  "maxProposition": 2,
  "name": "test",
  "semester": "S5",
  "startAt": 0,
  "updatedAt": Any<String>,
}
`;

exports[`POST /mentoringPropositions/:id/internships/:internship_type_id/link MentoringPropositions_200_WithInternship 1`] = `
Object {
  "additional": null,
  "address": "TEST",
  "city": "Brest",
  "country": "France",
  "createdAt": Any<String>,
  "description": "Stage",
  "endAt": 20191012,
  "id": 2,
  "isInternshipAbroad": true,
  "isProposition": false,
  "isPublish": true,
  "isValidated": false,
  "postalCode": "29280",
  "publishAt": 20191012,
  "startAt": 20191012,
  "state": "available",
  "subject": "test",
  "updatedAt": Any<String>,
}
`;

exports[`POST /mentoringPropositions/:id/mentors/:mentor_id/link MentoringPropositions_200_WithMentor 1`] = `
Object {
  "createdAt": Any<String>,
  "email": "john.doe@enib.fr",
  "firstName": "John",
  "id": 3,
  "lastName": "Doe",
  "role": "default",
  "updatedAt": Any<String>,
}
`;

exports[`PUT /mentoringPropositions/:id MentoringPropositions_200_NotAnyDataUpdated 1`] = `
Object {
  "comment": "test",
  "createdAt": Any<String>,
  "id": Any<Number>,
  "updatedAt": Any<String>,
}
`;

exports[`PUT /mentoringPropositions/:id MentoringPropositions_200_UpdateAllData 1`] = `
Object {
  "comment": "additional data",
  "createdAt": Any<String>,
  "id": Any<Number>,
  "updatedAt": Any<String>,
}
`;
